demos/saddle_point_pc/saddle_point_systems.py.rst

```diff
@@ -466,72 +460,61 @@ Instead of asking PETSc to build an approximation to :math:`S` which
+                - inner(jump(u, n), avg(grad(v)))*dS \
+                - inner(avg(grad(u)), jump(v, n), )*dS \
+                + alpha/h_avg * inner(jump(u, n), jump(v, n))*dS \
+                - inner(u*n, grad(v))*ds \
+                - inner(grad(u), v*n)*ds \
+                + (gamma/h)*inner(u, v)*ds)
+            bcs = None
+            return (a_dg, bcs)
+  
     parameters = {
         "ksp_type": "gmres",
         "ksp_rtol": 1e-8,
         "pc_type": "fieldsplit",
         "pc_fieldsplit_type": "schur",
         "pc_fieldsplit_schur_fact_type": "full",
-        "pc_fieldsplit_off_diag_use_amat": True,
         "fieldsplit_0_ksp_type": "preonly",
         "fieldsplit_0_pc_type": "ilu",
         "fieldsplit_1_ksp_type": "preonly",
-        "fieldsplit_1_pc_type": "hypre"
+        "fieldsplit_1_pc_type": "python",
+        "fieldsplit_1_pc_python_type": __name__+ ".DGLaplacian",
+        "fieldsplit_1_aux_pc_type": "hypre"
     }
-
-Additionally, we need to provide our ``build_problem`` function with
-the ability to construct this operator.  To do so, we define a
-function to pass as the ``aP`` argument. ::
-
-    def dg_laplacian(W):
-        sigma, u = TrialFunctions(W)
-        tau, v = TestFunctions(W)
-        n = FacetNormal(W.mesh())
-        alpha = Constant(4.0)
-        gamma = Constant(8.0)
-        h = CellSize(W.mesh())
-        h_avg = (h('+') + h('-'))/2
-        a_dg = dot(sigma, tau)*dx \
-               + dot(grad(v), grad(u))*dx \
-               - dot(avg(grad(v)), jump(u, n))*dS \
-               - dot(jump(v, n), avg(grad(u)))*dS \
-               + alpha/h_avg * dot(jump(v, n), jump(u, n))*dS \
-               - dot(grad(v), u*n)*ds \
-               - dot(v*n, grad(u))*ds \
-               + (gamma/h)*dot(v, u)*ds
-
-        return a_dg
-
-Now we just need to pass this extra argument to the ``build_problem``
-function ::
-
+    
     print("DG approximation for S_p")
     for n in range(8):
-        solver, w, b = build_problem(n, parameters, aP=dg_laplacian, block_matrix=True)
-        solver.solve(w, b)
-        print(w.function_space().mesh().num_cells(), solver.ksp.getIterationNumber())
+        solver, w = build_problem(n, parameters, aP=None, block_matrix=False)
+        solver.solve()
+        print(w.function_space().mesh().num_cells(), solver.snes.ksp.getIterationNumber())
 
 This actually results in slightly worse convergence than the diagonal
 approximation we used above.
 
 ============== ==================
  Mesh elements  GMRES iterations
 ============== ==================
-      2              2
-      8              10
-      32             17
-      128            20
-      512            19
-      2048           19
-      8192           18
-      32768          18
+    2                 2
+    8                 12
+    32                23
+    128               26
+    512               28
+    2048              26
+    8192              26
+    32768             24
```

Comment:
I got lost after all the back and forth here. Why do the iteration counts change?

Reply:
The Laplacian has to be negative, because the Schur complement is negative definite for this problem.